### PR TITLE
(SIMP-4953) enterprise-meta#98 Add SCE Version 2

### PR DESCRIPTION
**Pull request by heliocentric**
* Add version 2 file format, and create SCE (SIMP Compliance Engine) version 2.0
* Add support for telemetry data.
* Add functions for debugging purposes.

SIMP-4953 #close
Fixes enterprise-meta#198, enterprise-meta#356, enterprise-meta#136, enterprise-meta#98, enterprise-meta#97 enterprise-meta#324

### CODE REVIEW

**Comment by trevor-vaughan** [comment]
:heavy_check_mark: Existing `pam` module tests work properly with this patch.

File: simp/compliance_profiles/testdata.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+profiles:
+  disa:
+    checks:
+      disacheck: true
+  nist:
+    checks:
+      nistcheck: true
+checks:
+  nistcheck:
+    type: 'puppet-class-parameter'
```
Comment (trevor-vaughan) [inline simp/compliance_profiles/testdata.yaml]: Shouldn't this just be the default and/or discovered from the environment? I mean, we can inspect and see if we're running inside Puppet.

File: CHANGELOG [hunk 2]
```diff
@@ -1,3 +1,19 @@
...
+  - Add compliance_markup::telemetry() function to return filenames used to
+    load compliance data
+  - Add compliance_markup::debug::dump hiera parameter that dumps the cached
+    result hash for hiera debugging.
+  - Remove pupmod::ca_crl_pull_interval from nist_800_53_rev4.json's since
+    the parameter is deprecated.
+  - refactor loader to look in:
+    - "/SIMP/compliance_profiles"
+    - "/simp/compliance_profiles"
+    - "/lib/puppetx/compliance"
```
Comment (trevor-vaughan) [inline CHANGELOG]: This should probably be `lib/simp` and we really shouldn't use this unless we want all of the data to get shipped to the clients.

File: simp-compliance-engine.gemspec [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  s.description = 'A library for accessing the SIMP metadata format for the simp project'
+  s.authors = [
+    "SIMP Project"
+  ]
+  s.executables      = `git ls-files -- exe/*`.split("\n").map{ |f| File.basename(f) }
+  s.bindir           = 'exe'
+  s.email = 'simp@simp-project.org'
+  s.license = 'Apache-2.0'
+  s.homepage = 'https://github.com/simp/pupmod-simp-compliance_markup'
+  s.files = Dir['Rakefile', '{bin,lib,spec}/**/*', 'README*','*.gemspec','*.md','CONTRIBUTORS','Gemfile','LICENSE','NOTICE.TXT', '.gem_version', '.gem_date']
```
[the remark this answers is not in the document]
Comment (heliocentric) [reply simp-compliance-engine.gemspec]: fix gemspec. Gemspec is still needed as downstream consumers pull in the library as a gem dependency.

File: simp/compliance_profiles/test.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+require 'yaml'
+Dir.glob("profile*.yaml") do |filename|
+	data = YAML.load(File.read(filename))
```
[the remark this answers is not in the document]
Comment (heliocentric) [reply simp/compliance_profiles/test.rb]: oh, this wasn't supposed to be in this commit

File: simp/compliance_profiles/testdata.yaml (hunk 1 above)
Comment (heliocentric) [reply simp/compliance_profiles/testdata.yaml]: no, every check is explicitly typed

**Comment by kendall-moore** [comment]
@op-ct I want to address a running theme from your review centered around testing. Namely, that this PR contains too few tests to be able to be merged. I agree that testing is infinitely important, and that assurances can be made with well-written tests, but I think that if we are going to block PRs from being accepted it stands to reason that we ought to have a testing threshold to measure PRs against. In order to keep this relevant to only this PR - what is that threshold here? It might help to both @heliocentric and anyone else hoping to hop on with getting this accepted to know where that is so that we can start adding tests with a specific goal in mind.

**Comment by cmentzer** [comment]
@ayohrling @kendall-moore To be honest, I'm actually ok with a relatively low amount of automated or formalized tests, as long as the ones responsible for the code have done manual testing and are confident in their work. I'd Imagine that Dylan has done quite a bit over the course of writing the code to ensure that it works and is valid, and while formalized tests are important we might just not have the number of man-hours we need to get full coverage. 

What resonates with me is that I wish it was easier to come into the code from the outside and have a good understanding of the logic behind the decisions being made, both at a high level (maybe link to wiki pages, or meeting notes) and a low level (comments NEED to exist in line with the code that is written). 

That being said I've skimmed the changes and I'm not seeing anything obvious that would give me concerns about functionality. It's a lot of code, and it's unrealistic to expect to be able to jump into it and know if it's good. @heliocentric put a lot of effort into it and if he's confident that it's ready to be merged and none of us see obvious concerns then he's probably right.

**Comment by trevor-vaughan** [comment]
I've mapped over the PAM compliance data at https://github.com/trevor-vaughan/pupmod-simp-pam/tree/compliance-2-testing but I have not been able to determine how to properly enable the backend for acceptance testing.

@heliocentric Any guidance on how to add the acceptance tests would be appreciated

@ayohrling and @cmentzer I'll be working to add the usual expected testing procedures, as well as guidance for what we've generally been following as best practice. I just realized that we haven't gotten that up to date. I'll ping you for review when it's ready.

**Comment by trevor-vaughan** [comment]
@ayohrling and @cmentzer I made the documentation updates over the weekend to reflect the current process and they're up for PR/discussion at https://github.com/simp/simp-doc/pull/241

**Comment by trevor-vaughan** [comment]
The compliance tests at https://github.com/trevor-vaughan/pupmod-simp-pam/tree/compliance-2-testing were successful in comparison to the previous version.

**Comment by trevor-vaughan** [comment]
Merged based on prior reviews and tests in the PAM branch referenced above.

**Pull request merged**